diff --git a/include/openssl/prefix_symbols.h b/include/openssl/prefix_symbols.h
index d4c1076..421fa33 100644
--- a/include/openssl/prefix_symbols.h
+++ b/include/openssl/prefix_symbols.h
@@ -3152,6 +3152,7 @@
 #pragma redefine_extname SSL_CTX_set_delegated_credentials BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_set_delegated_credentials)
 #pragma redefine_extname SSL_CTX_set_extension_order BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_set_extension_order)
 #pragma redefine_extname SSL_CTX_set_record_size_limit BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_set_record_size_limit)
+#pragma redefine_extname SSL_CTX_set_grease_signature_algorithms BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_set_grease_signature_algorithms)
 #pragma redefine_extname SSL_CTX_set_tls13_cipher_order BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_set_tls13_cipher_order)
 #pragma redefine_extname SSL_set_record_size_limit BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_set_record_size_limit)
 #pragma redefine_extname SSL_set_tls12_extensions BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_set_tls12_extensions)
@@ -6273,6 +6274,7 @@
 #define SSL_CTX_set_delegated_credentials BORINGSSL_ADD_PREFIX(SSL_CTX_set_delegated_credentials)
 #define SSL_CTX_set_extension_order BORINGSSL_ADD_PREFIX(SSL_CTX_set_extension_order)
 #define SSL_CTX_set_record_size_limit BORINGSSL_ADD_PREFIX(SSL_CTX_set_record_size_limit)
+#define SSL_CTX_set_grease_signature_algorithms BORINGSSL_ADD_PREFIX(SSL_CTX_set_grease_signature_algorithms)
 #define SSL_CTX_set_tls13_cipher_order BORINGSSL_ADD_PREFIX(SSL_CTX_set_tls13_cipher_order)
 #define SSL_set_record_size_limit BORINGSSL_ADD_PREFIX(SSL_set_record_size_limit)
 #define SSL_set_tls12_extensions BORINGSSL_ADD_PREFIX(SSL_set_tls12_extensions)
diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index 8ca1d0c..fc224fd 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -5551,6 +5551,9 @@ OPENSSL_EXPORT int SSL_CTX_set_tls13_cipher_order(SSL_CTX *ctx,
                                                   const uint16_t *order,
                                                   size_t order_len);
 
+OPENSSL_EXPORT void SSL_CTX_set_grease_signature_algorithms(SSL_CTX *ctx,
+                                                            int enabled);
+
 // SSL_set_permute_extensions configures whether sockets on `ssl` should
 // permute extensions. For now, this is only implemented for the ClientHello.
 OPENSSL_EXPORT void SSL_set_permute_extensions(SSL *ssl, int enabled);
diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index e594ec2..cdd3f61 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -1032,6 +1032,9 @@ static bool ext_sigalgs_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
   if (!CBB_add_u16(out_compressible, TLSEXT_TYPE_signature_algorithms) ||
       !CBB_add_u16_length_prefixed(out_compressible, &contents) ||
       !CBB_add_u16_length_prefixed(&contents, &sigalgs_cbb) ||
+      (hs->ssl->ctx->grease_signature_algorithms &&
+       !CBB_add_u16(&sigalgs_cbb, ssl_get_grease_value(
+                                      hs, ssl_grease_signature_algorithm))) ||
       !tls12_add_verify_sigalgs(hs, &sigalgs_cbb) ||
       !CBB_flush(out_compressible)) {
     return false;
diff --git a/ssl/internal.h b/ssl/internal.h
index e2ff882..32ea4dd 100644
--- a/ssl/internal.h
+++ b/ssl/internal.h
@@ -1667,7 +1667,8 @@ enum ssl_grease_index_t {
   ssl_grease_version,
   ssl_grease_ticket_extension,
   ssl_grease_ech_config_id,
-  ssl_grease_last_index = ssl_grease_ech_config_id,
+  ssl_grease_signature_algorithm,
+  ssl_grease_last_index = ssl_grease_signature_algorithm,
 };
 
 enum tls12_server_hs_state_t {
@@ -4200,6 +4201,8 @@ class SSLContext : public ssl_ctx_st, public RefCounted<SSLContext> {
   // grease_enabled is whether GREASE (RFC 8701) is enabled.
   bool grease_enabled : 1;
 
+  bool grease_signature_algorithms : 1;
+
   // permute_extensions is whether to permute extensions when sending messages.
   bool permute_extensions : 1;
 
diff --git a/ssl/ssl_lib.cc b/ssl/ssl_lib.cc
index 94d311e..5f367bb 100644
--- a/ssl/ssl_lib.cc
+++ b/ssl/ssl_lib.cc
@@ -388,6 +388,7 @@ bssl::SSLContext::SSLContext(const SSL_METHOD *ssl_method)
       signed_cert_timestamps_enabled(false),
       channel_id_enabled(false),
       grease_enabled(false),
+      grease_signature_algorithms(false),
       permute_extensions(false),
       allow_unknown_alpn_protos(false),
       false_start_allowed_without_alpn(false),
@@ -3187,6 +3188,10 @@ void SSL_CTX_set_grease_enabled(SSL_CTX *ctx, int enabled) {
   FromOpaque(ctx)->grease_enabled = !!enabled;
 }
 
+void SSL_CTX_set_grease_signature_algorithms(SSL_CTX *ctx, int enabled) {
+  FromOpaque(ctx)->grease_signature_algorithms = !!enabled;
+}
+
 void SSL_CTX_set_permute_extensions(SSL_CTX *ctx, int enabled) {
   FromOpaque(ctx)->permute_extensions = !!enabled;
 }
